src/Sylius/Component/Core/Specification/Specification.php

```diff
@@ -0,0 +1,21 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Core\Specification;
+
+interface Specification
+{
+    public function isSatisfiedBy(object $candidate): bool;
+
+    public function and(Specification $specification): Specification;
```

Comment:
The Specification should only require `isSatisfiedBy` and the composition should be handled by AndSpecification, OrSpecification, etc.
That's why you had to add the CompositeSpecification with an implementation of `and`.